src/main/java/seedu/todo/ui/InputHandler.java

```diff
@@ -1,12 +1,95 @@
 package seedu.todo.ui;
 
+import java.util.LinkedList;
+
 import seedu.todo.controllers.*;
 
 public class InputHandler {
+    
+    private static InputHandler instance;
+    
+    private static final int MAX_HISTORY_SIZE = 20;
+    private static final int INITIAL_HISTORY_INDEX = -1;
+    private static LinkedList<String> commandHistory = new LinkedList<String>();
+    private static int commandHistoryCurrentIndex = INITIAL_HISTORY_INDEX;
+
+    private Controller handlingController = null;
+    
+    protected InputHandler() {
+        // Prevent instantiation.
+    }
+    
+    /**
+     * Gets the current input handler instance.
+     */
+    public static InputHandler getInstance() {
+        if (instance == null) {
+            instance = new InputHandler();
+        }
+        
+        return instance;
+    }
+    
+    /**
+     * Pushes a command to the start of a LinkedList.
+     * @param command   Command string
+     */
+    private void pushCommand(String command) {
+        // Resets current pointer.
+        commandHistoryCurrentIndex = INITIAL_HISTORY_INDEX;
+        
+        // Adds to the start of the LinkedList.
+        commandHistory.addFirst(command);
+        
+        if (commandHistory.size() > MAX_HISTORY_SIZE) {
+            commandHistory.removeLast();
+        }
+    }
+    
+    /**
+     * Gets the previous command from the command history. Successive calls will return commands earlier in history.
+     * 
+     * @return  The input command earlier than what was previously retrieved
+     */
+    public String getPreviousCommandFromHistory() {
+        // Advance index only up to commandHistory.size()
+        // Advances down the list
+        if (commandHistoryCurrentIndex < commandHistory.size()) {
+            commandHistoryCurrentIndex++;
+        }
+        
+        // Returns empty String for index out of bounds
+        if (commandHistoryCurrentIndex >= commandHistory.size()) {
+            return "";
+        } else {
+            return commandHistory.get(commandHistoryCurrentIndex);
+        }
+    }
+    
+    /**
+     * Gets the next command from the command history. Successive calls will return commands later in history.
+     * 
+     * @return  The input command later than what was previously retrieved
+     */
+    public String getNextCommandFromHistory() {
+        // Advance index only up to INITIAL_HISTORY_INDEX
+        // Advance up the list
+        if (commandHistoryCurrentIndex - 1 >= INITIAL_HISTORY_INDEX) {
+            commandHistoryCurrentIndex--;
+        }
 
-    Controller handlingController = null;
+        // Returns empty String for index out of bounds
+        if (commandHistoryCurrentIndex < 0) {
+            return "";
+        } else {
+            return commandHistory.get(commandHistoryCurrentIndex);
+        }
+    }
 
     public boolean processInput(String input) {
+        // Push to history
+        pushCommand(input);
```

Comment:
Should we really be saving the command even if the "false" case happens below (i.e. we cannot find a controller to process the command)? Since the command is 1. wrong 2. stays unchanged on screen (not even disambiguated), I think we should consider skipping saving the wrong command.